pulsar-protocols/grpc/src/main/proto/PulsarApi.proto

```diff
@@ -0,0 +1,888 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+syntax = "proto2";
+
+package pulsar.protocols.grpc;
+option java_package = "org.apache.pulsar.protocols.grpc.api";
+option java_multiple_files = true;
+// option optimize_for = LITE_RUNTIME;
+
+service Pulsar {
+    rpc produce(stream CommandSend) returns (stream SendResult) {}
```

Comment:
note that stream-based `produce` will not be using any load balancing and send messages to the same broker, while single record produce can use gRPC's load balancing underneath.

That's exactly the reason why we had it as a single record produce in Liiklus:
https://github.com/bsideup/liiklus/blob/e5620a2320811997fe8d09b62db078fdfed9acd4/protocol/src/main/proto/LiiklusService.proto#L13

Reply:
It would be very helpful to benchmark, because stream<->stream channel's overhead may be bigger than send-receive operation.
The underlying gRPC protocol is stateful and does framing and stuff, which kinda gives you the streaming semantics already.

Reply:
Produce messages need to be sent to the leader/topic owner broker only so you can't really do load balancing. Also this would require sharing the auth session (when it's stateful) between the brokers.
As for req/resp vs bidi-streaming, it can be benchmarked but the cost of authentication/authorization is probably a killer without some persistent session mechanism.
Now it shouldn't be hard to expose a req/resp endpoint in parallel of the bidi-streaming one. That can be easier to use for simple producers that just want to send a single message and be gone.

Reply:
> Produce messages need to be sent to the leader/topic owner broker only so you can't really do load balancing.

Since I foresee gRPC support in pulsar-proxy, this statement can be relaxed, I guess :)

> As for req/resp vs bidi-streaming, it can be benchmarked but the cost of authentication/authorization is probably a killer without some persistent session mechanism.

AFAIK auth is per gRPC session, not per stream. Once you establish it, req/resp messages go through the same channel and do not require any additional auth check (unless you have it in the logical protocol)

Reply:
> Now it shouldn't be hard to expose a req/resp endpoint in parallel of the bidi-streaming one. 

👍

Reply:
> Since I foresee gRPC support in pulsar-proxy, this statement can be relaxed, I guess :)

This PR is for using gRPC in place of the binary protocol. So without going through a proxy. There's another PR for a gRPC-proxy (#3032). I just need time to finish it.

> AFAIK auth is per gRPC session, not per stream. Once you establish it, req/resp messages go through the same channel and do not require any additional auth check (unless you have it in the logical protocol)

In Java, the only per-channel auth available is with mutual TLS. For other types of auth, you need to do per-call auth with an interceptor.

Reply:
>  So without going through a proxy. 

I understand this, but... does it mean that proxy and the broker will have different protocols? I expected them to be the same, so that the proxy becomes transparent (as it is with Pulsar's protocol)

Reply:
We could probably do the transparent gRPC proxy after the gRPC handler is done. The idea of the #3032 is more to have a proxy like the websocket proxy which simplifies a lot of things and uses the Java driver with binary protocol to communicate with the brokers.